### PR TITLE
Feature/tecg 296 follow technology

### PR DESCRIPTION
**Pull request by brunoliberal**

### CODE REVIEW

File: src/main/java/com/ciandt/techgallery/service/impl/CronServiceImpl.java [hunk 1]
```diff
@@ -66,13 +76,22 @@ public void sendEmailtoFollowers() {
...
+          TechGalleryActivitiesTO techGalleryActivitiesTo = new TechGalleryActivitiesTO();
+          techGalleryActivitiesTo.setTimestamp(new Date());
+          techGalleryActivitiesTo.setFollower(follower);
+          techGalleryActivitiesTo.setAppName(Constants.APP_NAME);
+          List<TechnologyActivitiesTO> techActivitiesToList =
+              new ArrayList<TechnologyActivitiesTO>();
+
+          for (String id : follower.getFollowedTechnologyIds()) {
+            Technology technology = technologyDao.findById(id);
+            // Find the last executed cronJob time
```
Comment (brunocarneiro) [inline src/main/java/com/ciandt/techgallery/service/impl/CronServiceImpl.java]: @brunoliberal Take a look at this method...it is long and hard to understand. Also, you already defined that this block should be a method using this comment. Please refactor this part to findLastExecutedCronJob

File: src/main/java/com/ciandt/techgallery/service/impl/CronServiceImpl.java [hunk 2]
```diff
@@ -82,28 +101,60 @@ public void sendEmailtoFollowers() {
...
-                    .findAllRecommendationsIdsStartingFrom(lastCronJobExecDate);
-            String dailyCommentsIds =
-                technologyCommentDao.findAllCommentsIdsStartingFrom(lastCronJobExecDate);
-            List<Ref<TechGalleryUser>> followers = technologyFollowers.getFollowers();
-            if (!dailyRecommendationsIds.isEmpty() || !dailyCommentsIds.isEmpty()) {
-              for (Ref<TechGalleryUser> ref : followers) {
-                TechGalleryUser follower = ref.get();
-                emailService.push(follower, technology, dailyRecommendationsIds, dailyCommentsIds);
+
+            // Find new Activities in a technology
```
Comment (brunocarneiro) [inline src/main/java/com/ciandt/techgallery/service/impl/CronServiceImpl.java]: @brunoliberal Also move this to a new method findNewActivitiesInATechnology

File: src/main/java/com/ciandt/techgallery/service/impl/CronServiceImpl.java [hunk 3]
```diff
@@ -82,28 +101,60 @@ public void sendEmailtoFollowers() {
...
     } catch (Exception e) {
+      _LOG.info(e.getMessage());
       cronJob.setEndTimestamp(new Date());
       cronJob.setCronStatus(CronStatus.FAILURE);
       cronJob.setDescription(e.getMessage());
     }
     cronJobsDao.add(cronJob);
   }
+
+  private TechnologyActivitiesTO createTechnologyActivitiesTo(Technology technology,
```
Comment (brunocarneiro) [inline src/main/java/com/ciandt/techgallery/service/impl/CronServiceImpl.java]: @brunoliberal this should be in TechnologyActivitiesService

File: src/main/java/com/ciandt/techgallery/service/model/TechnologyActivitiesTO.java [hunk 4]
```diff
@@ -17,29 +18,30 @@
...
   public String getTechnologyLink() {
-    return technologyLink;
-  }
-
-  public void setTechnologyLink(String technologyLink) {
-    this.technologyLink = technologyLink;
+    String linkTechnology;
+    String queryString = "?id=" + technology.getId();
+    String environment = System.getProperty(Constants.RUNTIME_ENVIRONMENT_PROPERTY);
+    if (StringUtils.equals(Constants.PRODUCTION_PROPERTY, environment)) {
```
Comment (brunocarneiro) [inline src/main/java/com/ciandt/techgallery/service/model/TechnologyActivitiesTO.java]: @brunoliberal We already have a class to do that. Take a look at TechGalleryUtil

File: src/main/java/com/ciandt/techgallery/service/impl/CronServiceImpl.java [hunk 5]
```diff
@@ -82,28 +101,60 @@ public void sendEmailtoFollowers() {
...
-            String dailyCommentsIds =
-                technologyCommentDao.findAllCommentsIdsStartingFrom(lastCronJobExecDate);
-            List<Ref<TechGalleryUser>> followers = technologyFollowers.getFollowers();
-            if (!dailyRecommendationsIds.isEmpty() || !dailyCommentsIds.isEmpty()) {
-              for (Ref<TechGalleryUser> ref : followers) {
-                TechGalleryUser follower = ref.get();
-                emailService.push(follower, technology, dailyRecommendationsIds, dailyCommentsIds);
+
+            // Find new Activities in a technology
+            List<TechnologyRecommendation> dailyRecommendations =
```
Comment (brunocarneiro) [inline src/main/java/com/ciandt/techgallery/service/impl/CronServiceImpl.java]: @brunoliberal Also move this to a method called something like 'findActivitiesByTechnology' or 'findDailyRecommendationsByTechnology'

File: src/main/java/com/ciandt/techgallery/service/impl/CronServiceImpl.java [hunk 6]
```diff
@@ -82,28 +101,60 @@ public void sendEmailtoFollowers() {
...
+            if (dailyRecommendations != null || dailyComments != null) {
+              // Create a TO to each technology
+              TechnologyActivitiesTO techActivitiesTo =
+                  createTechnologyActivitiesTo(technology, dailyRecommendations, dailyComments);
+              techActivitiesToList.add(techActivitiesTo);
+            }
+          }
+          techGalleryActivitiesTo.setTechnologyActivitiesTo(techActivitiesToList);
+
+          EmailConfig email =
```
Comment (brunocarneiro) [inline src/main/java/com/ciandt/techgallery/service/impl/CronServiceImpl.java]: @brunoliberal put this block in a method in EmailService. Something like 'pushEmailConfig'

File: src/main/java/com/ciandt/techgallery/service/impl/CronServiceImpl.java (hunk 1 above)
Comment (brunoliberal) [inline src/main/java/com/ciandt/techgallery/service/impl/CronServiceImpl.java]: Done

File: src/main/java/com/ciandt/techgallery/service/impl/CronServiceImpl.java (hunk 2 above)
Comment (brunoliberal) [inline src/main/java/com/ciandt/techgallery/service/impl/CronServiceImpl.java]: done

File: src/main/java/com/ciandt/techgallery/service/impl/CronServiceImpl.java (hunk 3 above)
Comment (brunoliberal) [inline src/main/java/com/ciandt/techgallery/service/impl/CronServiceImpl.java]: Ok. I created a new service and put this method in there.

File: src/main/java/com/ciandt/techgallery/service/impl/CronServiceImpl.java (hunk 5 above)
Comment (brunoliberal) [inline src/main/java/com/ciandt/techgallery/service/impl/CronServiceImpl.java]: done that in the second comment

File: src/main/java/com/ciandt/techgallery/service/model/TechnologyActivitiesTO.java (hunk 4 above)
Comment (brunoliberal) [inline src/main/java/com/ciandt/techgallery/service/model/TechnologyActivitiesTO.java]: done. I only removed the version variable. cause the method get the full path to technology view page link

**Pull request merged**